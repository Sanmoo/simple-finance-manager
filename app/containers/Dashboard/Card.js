import styled from 'styled-components';
import { Card as CardBase } from '@material-ui/core';

const Card = styled(CardBase)`
  width: 95%;
  margin: 0 auto;
  margin-top: 20px;
`;

export default Card;
