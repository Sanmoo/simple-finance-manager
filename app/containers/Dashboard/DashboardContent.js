import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardContent, Typography } from '@material-ui/core';

export default function DashboardComponent({ totalExpenses, totalReceipt }) {
  return (
    <Card className="content">
      <CardContent>
        <Typography variant="h5" component="h2">
          Performance this month
        </Typography>
        <p>Total receipt: R$ {totalReceipt}</p>
        <p>Total expenses: R$ {totalExpenses}</p>
      </CardContent>
    </Card>
  );
}

DashboardComponent.propTypes = {
  totalExpenses: PropTypes.number.isRequired,
  totalReceipt: PropTypes.number.isRequired,
};
