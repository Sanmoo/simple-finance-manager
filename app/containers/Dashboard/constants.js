// ACTION TYPES
export const SAVE_KEY = 'sfm/Dashboard/SAVE_KEY';
export const CREATE_NEW_SPREADSHEET = 'sfm/Dashboard/CREATE_NEW_SPREADSHEET';
export const ON_SYNC_LOCAL_CACHE = 'sfm/Dashboard/ON_SYNC_LOCAL_CACHE';
export const CREATE_SHEET_FROM_GENERIC_TEMPLATE =
  'sfm/Dashboard/CREATE_SHEET_FROM_GENERIC_TEMPLATE';
export const CREATE_SHEET_FROM_PREVIOUS_MONTH =
  'sfm/Dashboard/CREATE_SHEET_FROM_PREVIOUS_MONTH';
export const TRY_SYNC_AGAIN = 'sfm/Dashboard/TRY_SYNC_AGAIN';

// OTHER
export const ACTIONABLE_DIALOG_NO_SHEETS = 'ACTIONABLE_DIALOG_NO_SHEETS';
export const ACTIONABLE_DIALOG_PREVIOUS_MONTH_SHEET =
  'ACTIONABLE_DIALOG_PREVIOUS_MONTH_SHEET';
