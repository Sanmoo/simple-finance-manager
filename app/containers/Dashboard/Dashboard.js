import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import SpreadsheetIdInputCard from 'components/SpreadsheetIdInputCard';
import { syncLocalCache, getDashboardInfo } from 'utils/business';
import { withStyles } from '@material-ui/core/styles';
import SpeedDial from '@material-ui/lab/SpeedDial';
import SpeedDialIcon from '@material-ui/lab/SpeedDialIcon';
import SpeedDialAction from '@material-ui/lab/SpeedDialAction';
import ReceiptIcon from '@material-ui/icons/Receipt';
import AttachMoneyIcon from '@material-ui/icons/AttachMoney';

import DashboardContent from './DashboardContent';

const styles = theme => ({
  speedDialWrapper: {
    position: 'relative',
  },
  speedDial: {
    position: 'fixed',
    bottom: theme.spacing.unit * 2,
    right: theme.spacing.unit * 3,
  },
});

const DashboardContainer = styled.div`
  padding-bottom: 56px;
  position: relative;
`;

const actions = [
  { icon: <AttachMoneyIcon />, name: 'Add Income' },
  { icon: <ReceiptIcon />, name: 'Add Expense' },
];

function Dashboard({
  sId,
  saveKey,
  dashInfo,
  classes,
  onAddExpense,
  onAddIncome,
  onSpreadsheetIdProvided,
  addEntryButtonEnabled,
  onCreateNewSpreadsheet,
  isLoading,
}) {
  useEffect(() => {
    if (sId) {
      syncLocalCache(sId)
        .then(getDashboardInfo)
        .then(value => saveKey('dashInfo', value));
    }
  }, [sId, saveKey]);

  const [open, setOpen] = useState(false);

  if (!sId) {
    return (
      <SpreadsheetIdInputCard
        onCreateNewSpreadsheet={onCreateNewSpreadsheet}
        onSubmitSId={onSpreadsheetIdProvided}
        isLoading={isLoading}
      />
    );
  }

  const onClose = () => setOpen(false);
  const onOpen = () => setOpen(true);
  const onClick = () => setOpen(!open);

  return (
    <DashboardContainer>
      <DashboardContent dashInfo={dashInfo} />
      <div className={classes.speedDialWrapper}>
        {addEntryButtonEnabled && (
          <SpeedDial
            ariaLabel="Action menu"
            className={classes.speedDial}
            hidden={false}
            icon={<SpeedDialIcon />}
            onBlur={onClose}
            onClick={onClick}
            onClose={onClose}
            onFocus={onOpen}
            onMouseEnter={onOpen}
            onMouseLeave={onClose}
            open={open}
            direction="up"
          >
            {actions.map(action => (
              <SpeedDialAction
                key={action.name}
                icon={action.icon}
                tooltipTitle={action.name}
                tooltipOpen
                onClick={() =>
                  action.name === 'Add Income' ? onAddIncome() : onAddExpense()
                }
              />
            ))}
          </SpeedDial>
        )}
      </div>
    </DashboardContainer>
  );
}

Dashboard.propTypes = {
  sId: PropTypes.string,
  saveKey: PropTypes.func.isRequired,
  onSpreadsheetIdProvided: PropTypes.func.isRequired,
  dashInfo: PropTypes.any,
  classes: PropTypes.object.isRequired,
  onAddIncome: PropTypes.func.isRequired,
  onAddExpense: PropTypes.func.isRequired,
  addEntryButtonEnabled: PropTypes.bool.isRequired,
  onCreateNewSpreadsheet: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default withStyles(styles)(Dashboard);
