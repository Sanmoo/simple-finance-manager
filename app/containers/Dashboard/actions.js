import { SAVE_KEY, CREATE_NEW_SPREADSHEET } from './constants';

export function saveKey(key, value) {
  return {
    type: SAVE_KEY,
    key,
    value,
  };
}

export function createNewSpreadsheet() {
  return { type: CREATE_NEW_SPREADSHEET };
}
