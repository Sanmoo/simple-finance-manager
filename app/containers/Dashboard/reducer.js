import produce from 'immer';
import { SIGN_OFF } from 'containers/App/constants';
import { SAVE_KEY } from './constants';

// The initial state of the App
export const initialState = {
  dashInfo: [],
  isLoading: false,
};

/* eslint-disable default-case, no-param-reassign, consistent-return */
const reducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case SAVE_KEY:
        draft[action.key] = action.value;
        break;
      case SIGN_OFF:
        return initialState;
    }
  });

export default reducer;
