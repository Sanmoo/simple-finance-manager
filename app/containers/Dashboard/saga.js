/**
 * Gets the repositories of the user from Github
 */

import { takeLatest, put } from 'redux-saga/effects';
import {
  SAVE_KEY,
  CREATE_NEW_SPREADSHEET,
  ON_SYNC_LOCAL_CACHE,
  ACTIONABLE_DIALOG_NO_SHEETS,
  ACTIONABLE_DIALOG_PREVIOUS_MONTH_SHEET,
  CREATE_SHEET_FROM_GENERIC_TEMPLATE,
  CREATE_SHEET_FROM_PREVIOUS_MONTH,
  TRY_SYNC_AGAIN,
} from 'containers/Dashboard/constants';
import {
  createNewSpreadsheetFromTemplate,
  createSheetFromGenericTemplate,
  createSheetFromPreviousMonth,
  getDashboardInfo,
  syncLocalCache,
  getPreviousMonthSheetName,
  verifySheetExistence,
  getSpreadsheetUrl,
} from 'utils/business';
import { saveKey as saveGlobalKey } from 'containers/App/actions';
import { ERROR_SHEET_NOT_FOUND } from 'utils/constants';
import { saveKey, onSyncLocalCache } from './actions';

export function* handleSaveKey({ key, value }) {
  if (key === 'spreadsheetId') {
    window.localStorage.setItem('spreadsheetId', value);
  }
}

export function* handleCreateNewSpreadsheet() {
  try {
    yield put(saveKey('isLoading', true));
    const { spreadsheetId } = yield createNewSpreadsheetFromTemplate();
    yield put(saveGlobalKey('spreadsheetId', spreadsheetId));
  } catch (e) {
    /* eslint-disable no-alert */
    alert('There was an error while trying to create the spreadsheet.');
  } finally {
    yield put(saveKey('isLoading', false));
  }
}

export function* handleSyncLocalCache({ sId }) {
  try {
    yield syncLocalCache(sId);
    const dashInfo = yield getDashboardInfo();
    const spreadsheetUrl = yield getSpreadsheetUrl(sId);
    yield put(saveKey('dashInfo', dashInfo));
    yield put(saveGlobalKey('spreadsheetUrl', spreadsheetUrl));
  } catch (e) {
    if (e && e.type === ERROR_SHEET_NOT_FOUND) {
      const lastMonthSheetName = getPreviousMonthSheetName();
      if (yield verifySheetExistence(sId, lastMonthSheetName)) {
        yield put(
          saveKey(
            'actionableDialogType',
            ACTIONABLE_DIALOG_PREVIOUS_MONTH_SHEET,
          ),
        );
      } else {
        yield put(saveKey('actionableDialogType', ACTIONABLE_DIALOG_NO_SHEETS));
      }
    } else {
      alert(
        'There was an unpexpected error while trying to sync with upstream spreadsheet',
      );
    }
  }
}

export function* handleCreateSheetFromGenericTemplate({ sId }) {
  yield put(saveKey('actionableDialogType', null));
  yield createSheetFromGenericTemplate(sId);
  yield put(onSyncLocalCache(sId));
}

export function* handleCreateSheetFromPreviousMonth({ sId }) {
  yield put(saveKey('actionableDialogType', null));
  yield createSheetFromPreviousMonth(sId);
  yield put(onSyncLocalCache(sId));
}

export function* handleTrySyncAgain({ sId }) {
  yield put(saveKey('actionableDialogType', null));
  yield put(onSyncLocalCache(sId));
}

export default function* saga() {
  yield takeLatest(SAVE_KEY, handleSaveKey);
  yield takeLatest(CREATE_NEW_SPREADSHEET, handleCreateNewSpreadsheet);
  yield takeLatest(ON_SYNC_LOCAL_CACHE, handleSyncLocalCache);
  yield takeLatest(
    CREATE_SHEET_FROM_GENERIC_TEMPLATE,
    handleCreateSheetFromGenericTemplate,
  );
  yield takeLatest(
    CREATE_SHEET_FROM_PREVIOUS_MONTH,
    handleCreateSheetFromPreviousMonth,
  );
  yield takeLatest(TRY_SYNC_AGAIN, handleTrySyncAgain);
}
