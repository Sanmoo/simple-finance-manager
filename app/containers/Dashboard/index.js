import { memo } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { push } from 'connected-react-router';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import { makeSelectSId } from 'containers/App/selectors';
import { saveKey as saveGlobalKey } from 'containers/App/actions';

import Dashboard from './Dashboard';
import { saveKey } from './actions';
import { makeSelectDashInfo } from './selectors';

import reducer from './reducer';
import saga from './saga';

const mapStateToProps = createStructuredSelector({
  sId: makeSelectSId(),
  dashInfo: makeSelectDashInfo(),
});

export function mapDispatchToProps(dispatch) {
  return {
    saveKey: (key, val) => dispatch(saveKey(key, val)),
    onAddClick: () => dispatch(push('/entry')),
    onSpreadsheetIdProvided: sId =>
      dispatch(saveGlobalKey('spreadsheetId', sId)),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({ key: 'dashboard', reducer });
const withSaga = injectSaga({ key: 'dashboard', saga });

export default compose(
  withConnect,
  withReducer,
  withSaga,
  memo,
)(Dashboard);
