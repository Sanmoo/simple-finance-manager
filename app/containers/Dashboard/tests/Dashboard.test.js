/**
 *
 * Tests for EditEntryPage
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import 'fake-indexeddb/auto';

import { Dashboard } from '../Dashboard';
import { DEFAULT_LOCALE } from '../../../i18n';

describe('<Dashboard />', () => {
  let props;

  beforeEach(() => {
    props = {
      sId: 'mysid',
      onSpreadsheetIdProvided: jest.fn(),
      dashInfo: [],
      onAddIncome: jest.fn(),
      onAddExpense: jest.fn(),
      addEntryButtonEnabled: true,
      onCreateNewSpreadsheet: jest.fn(),
      isLoading: false,
      actionableDialogType: 'type',
      onCreateSheetFromGenericTemplate: jest.fn(),
      onTrySyncAgain: jest.fn(),
      onCreateSheetFromPreviousMonth: jest.fn(),
      classes: {},
    };
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Dashboard {...props} />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Dashboard {...props} />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  describe('when sId is not provided', () => {
    beforeEach(() => {
      props.sId = null;
    });

    it('should render input component and match the snapshot', () => {
      const {
        container: { firstChild },
        container,
      } = render(
        <IntlProvider locale={DEFAULT_LOCALE}>
          <Dashboard {...props} />
        </IntlProvider>,
      );
      expect(container.querySelector('input')).not.toBeNull();
      expect(firstChild).toMatchSnapshot();
    });
  });

  describe('when sId is provided', () => {
    beforeEach(() => {
      props.sId = 'something';
    });

    it('should not render input component and match the snapshot', () => {
      const {
        container: { firstChild },
        container,
      } = render(
        <IntlProvider locale={DEFAULT_LOCALE}>
          <Dashboard {...props} />
        </IntlProvider>,
      );
      expect(container.querySelector('input')).toBeNull();
      expect(firstChild).toMatchSnapshot();
    });
  });
});
