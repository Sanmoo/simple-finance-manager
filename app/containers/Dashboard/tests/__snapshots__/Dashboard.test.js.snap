// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> Should render and match the snapshot 1`] = `
<div
  class="Dashboard__DashboardContainer-u5wi3k-0 focEzc"
>
  <div
    class="MuiPaper-root-146 MuiPaper-elevation1-149 MuiPaper-rounded-147 MuiCard-root-145 content"
  >
    <div
      class="MuiCardContent-root-173"
    >
      <h2
        class="MuiTypography-root-174 MuiTypography-h5-190"
      >
        Goals and Expenses
      </h2>
      <div
        class="MuiPaper-root-146 MuiPaper-elevation2-150 MuiPaper-rounded-147 DashboardContent-root-143"
      >
        <table
          class="MuiTable-root-210"
        >
          <tbody
            class="MuiTableBody-root-211"
          />
        </table>
      </div>
    </div>
  </div>
  <div />
  <div />
  <div
    class="Dashboard-speedDialWrapper-141"
  >
    <div
      class="MuiSpeedDial-root-228 MuiSpeedDial-directionUp-230 Dashboard-speedDial-142"
    >
      <button
        aria-controls="Actionmenu-actions"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Action menu"
        class="MuiButtonBase-root-246 MuiFab-root-236 MuiFab-primary-238 MuiSpeedDial-fab-229"
        style="transform: scale(1); webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label-237"
        >
          <span
            class="MuiSpeedDialIcon-root-249"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-255 MuiSpeedDialIcon-icon-250"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-274"
        />
      </button>
      <div
        aria-orientation="vertical"
        class="MuiSpeedDial-actions-234 MuiSpeedDial-actionsClosed-235 MuiSpeedDial-directionUp-230"
        id="Actionmenu-actions"
        role="menu"
      >
        <button
          class="MuiButtonBase-root-246 MuiFab-root-236 MuiFab-sizeSmall-244 MuiSpeedDialAction-button-264 MuiSpeedDialAction-buttonClosed-265"
          role="menuitem"
          style="transition-delay: 60ms;"
          tabindex="-1"
          title="Add Income"
          type="button"
        >
          <span
            class="MuiFab-label-237"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-255"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-274"
          />
        </button>
        <button
          class="MuiButtonBase-root-246 MuiFab-root-236 MuiFab-sizeSmall-244 MuiSpeedDialAction-button-264 MuiSpeedDialAction-buttonClosed-265"
          role="menuitem"
          style="transition-delay: 30ms;"
          tabindex="-1"
          title="Add Expense"
          type="button"
        >
          <span
            class="MuiFab-label-237"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-255"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 17H6v-2h12v2zm0-4H6v-2h12v2zm0-4H6V7h12v2zM3 22l1.5-1.5L6 22l1.5-1.5L9 22l1.5-1.5L12 22l1.5-1.5L15 22l1.5-1.5L18 22l1.5-1.5L21 22V2l-1.5 1.5L18 2l-1.5 1.5L15 2l-1.5 1.5L12 2l-1.5 1.5L9 2 7.5 3.5 6 2 4.5 3.5 3 2v20z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-274"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Dashboard /> when sId is not provided should render input component and match the snapshot 1`] = `
<div
  class="MuiPaper-root-284 MuiPaper-elevation1-287 MuiPaper-rounded-285 MuiCard-root-283"
>
  <div
    class="MuiCardContent-root-311"
  >
    <h2
      class="MuiTypography-root-312 MuiTypography-h4-327"
    >
      Please provide your spreadsheet Id
    </h2>
    <div
      class="MuiFormControl-root-348 MuiFormControl-marginNormal-349 SpreadsheetIdInputCard__TextField-ocjwh8-0 ihlGeT"
    >
      <label
        class="MuiFormLabel-root-363 MuiInputLabel-root-352 MuiInputLabel-formControl-357 MuiInputLabel-animated-360"
        data-shrink="false"
        for="spreadsheetId"
      >
        Your spreadsheet Id
      </label>
      <div
        class="MuiInputBase-root-383 MuiInput-root-370 MuiInput-underline-374 MuiInputBase-formControl-384 MuiInput-formControl-371"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input-393 MuiInput-input-378"
          id="spreadsheetId"
          type="text"
          value=""
        />
      </div>
    </div>
    <p
      class="MuiTypography-root-312 MuiTypography-subheading-319"
    >
      Or click 'Create one for me' if you don't have one yet
    </p>
  </div>
  <div
    class="MuiCardActions-root-400"
  >
    <button
      class="MuiButtonBase-root-429 MuiButtonBase-disabled-430 MuiButton-root-403 MuiButton-text-405 MuiButton-flat-408 MuiButton-sizeSmall-426 MuiButton-disabled-423 MuiCardActions-action-402"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label-404"
      >
        OK
      </span>
    </button>
    <button
      class="MuiButtonBase-root-429 MuiButton-root-403 MuiButton-text-405 MuiButton-flat-408 MuiButton-sizeSmall-426 MuiCardActions-action-402"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-404"
      >
        Create one for me
      </span>
      <span
        class="MuiTouchRipple-root-432"
      />
    </button>
  </div>
</div>
`;

exports[`<Dashboard /> when sId is provided should not render input component and match the snapshot 1`] = `
<div
  class="Dashboard__DashboardContainer-u5wi3k-0 focEzc"
>
  <div
    class="MuiPaper-root-444 MuiPaper-elevation1-447 MuiPaper-rounded-445 MuiCard-root-443 content"
  >
    <div
      class="MuiCardContent-root-471"
    >
      <h2
        class="MuiTypography-root-472 MuiTypography-h5-488"
      >
        Goals and Expenses
      </h2>
      <div
        class="MuiPaper-root-444 MuiPaper-elevation2-448 MuiPaper-rounded-445 DashboardContent-root-441"
      >
        <table
          class="MuiTable-root-508"
        >
          <tbody
            class="MuiTableBody-root-509"
          />
        </table>
      </div>
    </div>
  </div>
  <div />
  <div />
  <div
    class="Dashboard-speedDialWrapper-439"
  >
    <div
      class="MuiSpeedDial-root-526 MuiSpeedDial-directionUp-528 Dashboard-speedDial-440"
    >
      <button
        aria-controls="Actionmenu-actions"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Action menu"
        class="MuiButtonBase-root-544 MuiFab-root-534 MuiFab-primary-536 MuiSpeedDial-fab-527"
        style="transform: scale(1); webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label-535"
        >
          <span
            class="MuiSpeedDialIcon-root-547"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-553 MuiSpeedDialIcon-icon-548"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-572"
        />
      </button>
      <div
        aria-orientation="vertical"
        class="MuiSpeedDial-actions-532 MuiSpeedDial-actionsClosed-533 MuiSpeedDial-directionUp-528"
        id="Actionmenu-actions"
        role="menu"
      >
        <button
          class="MuiButtonBase-root-544 MuiFab-root-534 MuiFab-sizeSmall-542 MuiSpeedDialAction-button-562 MuiSpeedDialAction-buttonClosed-563"
          role="menuitem"
          style="transition-delay: 60ms;"
          tabindex="-1"
          title="Add Income"
          type="button"
        >
          <span
            class="MuiFab-label-535"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-553"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-572"
          />
        </button>
        <button
          class="MuiButtonBase-root-544 MuiFab-root-534 MuiFab-sizeSmall-542 MuiSpeedDialAction-button-562 MuiSpeedDialAction-buttonClosed-563"
          role="menuitem"
          style="transition-delay: 30ms;"
          tabindex="-1"
          title="Add Expense"
          type="button"
        >
          <span
            class="MuiFab-label-535"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-553"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 17H6v-2h12v2zm0-4H6v-2h12v2zm0-4H6V7h12v2zM3 22l1.5-1.5L6 22l1.5-1.5L9 22l1.5-1.5L12 22l1.5-1.5L15 22l1.5-1.5L18 22l1.5-1.5L21 22V2l-1.5 1.5L18 2l-1.5 1.5L15 2l-1.5 1.5L12 2l-1.5 1.5L9 2 7.5 3.5 6 2 4.5 3.5 3 2v20z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-572"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
